### ISSUE TITLE
Smart card reader is not being ignored by `--list`, `--list-all` when no card inserted

### ISSUE THREAD

**Issue by kai-tub** [str4d/age-plugin-yubikey#81]
## What were you trying to do
I was trying to use `age-plugin-yubikey`.

### Short
I am getting the following error:

```
age-plugin-yubikey --list-all

Error: Error while communicating with YubiKey: PC/SC error
Cause: The operation requires a Smart Card, but no Smart Card is currently in the device

[ Did this not do what you expected? Could an error be more useful? ]
[ Tell us: https://str4d.xyz/age-plugin-yubikey/report              ]
```

and I don't know where to go from here.
It would be nice to have a beginner's guide section for people that aren't familiar with PIV and get here because they know age and happen to have a yubikey laying around for other things (FIDO for example)

### Longer
Thank you for developing this plugin!
I got here because I am familiarizing myself with age/rage and using it for various tasks.
As I also own a yubikey, I thought it would be interesting to combine it with age and check this plugin out.
But, I am not at all familiar with PIV/smart cards, and it is a bit daunting to understand how to put the pieces together.
I think that yubico could make more effort to make it easier to introduce the PIV features, but I also guess that most people know what they are doing when getting a key as a smart card replacement.
So maybe I am just an odd duck in this case :D 

It would be helpful if the README could link to relevant resources to familiarize beginners with PIV and that additional functionality.
Especially a way to test that everything _else_ works before starting `age-plugin-yubikey`, as I assume that I have some other issues and that it isn't the plugin's fault.
That is my main request for the issue :) 

To provide some additional information for my error message: If I run `pcsc_scan` the yubikey is detected and identified as a smart card.

**Comment by str4d** [comment]
I suspect that the problem you are encountering is unrelated to your YubiKey, but being caused by a Smart Card reader. Does your laptop / device have a separate smart card reader or slot? Could you run `RUST_LOG=trace age-plugin-yubikey --list`?

**Comment by kai-tub** [comment]
Sorry, I installed a different OS on my laptop and it took some time to set everything back up.
Yes, I do have a separate smart card reader (slot).
This is the output:

```
[INFO  i18n_embed::requester] Current Locale: [LanguageIdentifier { language: Language(Some("en")), script: None, region: Some(Region("US")), variants: None }]
[DEBUG i18n_embed] Selecting translations for domain "age_plugin_yubikey"
[DEBUG i18n_embed] Searching for available languages, found language file: "en-US/age_plugin_yubikey.ftl"
[DEBUG i18n_embed] Requested Languages: [LanguageIdentifier { language: Language(Some("en")), script: None, region: Some(Region("US")), variants: None }]
[DEBUG i18n_embed] Available Languages: [LanguageIdentifier { language: Language(Some("en")), script: None, region: Some(Region("US")), variants: None }]
[DEBUG i18n_embed] Supported Languages: [LanguageIdentifier { language: Language(Some("en")), script: None, region: Some(Region("US")), variants: None }]
[DEBUG i18n_embed] Attempting to load language file: "en-US/age_plugin_yubikey.ftl"
[DEBUG i18n_embed::fluent] Loaded language file: "en-US/age_plugin_yubikey.ftl" for language: "en-US"
[ERROR yubikey::yubikey] error connecting to reader 'Alcor Micro AU9540 00 00': PC/SC error
[ERROR yubikey::yubikey] error connecting to reader 'Alcor Micro AU9540 00 00': PC/SC error
Error: Error while communicating with YubiKey: PC/SC error
Cause: The operation requires a Smart Card, but no Smart Card is currently in the device

[ Did this not do what you expected? Could an error be more useful? ]
[ Tell us: https://str4d.xyz/age-plugin-yubikey/report              ]
```
And with the excerpt from `lsusb`:
```
Bus 001 Device 007: ID 1050:0407 Yubico.com Yubikey 4/5 OTP+U2F+CCID
Bus 001 Device 005: ID 058f:9540 Alcor Micro Corp. AU9540 Smartcard Reader
```
It seems like it is trying to connect to the wrong device.

**Comment by str4d** [comment]
Interesting. I thought I was catching all of those errors, but I'm clearly missing a path.

In the meantime, you should be able to skip over that smart card reader by adding `--serial SERIAL_OF_YUBIKEY` to your command arguments. `SERIAL_OF_YUBIKEY` is the number printed on the underside of your YubiKey (e.g. `8765432`)

**Comment by kai-tub** [comment]
Ok, I can confirm that it works by setting `--serial` with the `--generate` command.
The key still had the default PIN, and when I tried to create a new PIN I was entering a PIN with 5 numbers and just got 
```
  Your YubiKey is using the default PIN. Let's change it!
✨ We'll also set the PUK equal to the PIN.

🔐 The PIN is up to 8 numbers, letters, or symbols. Not just numbers!
❌ Your keys will be lost if the PIN and PUK are locked after 3 incorrect tries.

Enter current PUK (default is 12345678): [hidden]
Choose a new PIN/PUK: [hidden]
[TRACE yubikey::transaction] going to send 16 bytes in this go
[TRACE yubikey::apdu] >>> Apdu { cla: 0, ins: ChangeReference, p1: 0, p2: 129, data: [49, 50, 51, 52, 53, 54, 55, 56, 50, 55, 49, 51, 57, 255, 255, 255] }
[TRACE yubikey::transaction] >>> [0, 36, 0, 129, 16, 49, 50, 51, 52, 53, 54, 55, 56, 50, 55, 49, 51, 57, 255, 255, 255]
[TRACE yubikey::apdu] <<< Response { status_words: IncorrectParamError, data: [] }
[ERROR yubikey::transaction] failed changing pin, token response code: 6a80.
Error: Error while communicating with YubiKey: generic error
```
It took me a moment to realize that I needed 6 symbols.
Maybe it would be possible to extend the message to
`🔐 The PIN can be between 6-8 numbers, letters, or symbols. Not just numbers!`

Although one can quickly find it from the yubico website: https://developers.yubico.com/yubikey-piv-manager/PIN_and_Management_Key.html I would find it better also to have the minimum number of characters included in the visual output and maybe also in the error message?

Thanks!

**Comment by str4d** [comment]
> Interesting. I thought I was catching all of those errors, but I'm clearly missing a path.

Aha, I've figured out the problem. I _did_ test against removed SmartCards when I implemented https://github.com/str4d/age-plugin-yubikey/commit/50b873c89ff57d03d98f70cc14c56331b43df79a using my Thinkpad's SmartCard reader (which resulted in `pcsc::Error::RemovedCard`), but the above error message corresponds to `pcsc::Error::NoSmartcard`.

I've opened #117 to fix this. Could you test it @kai-tub? If it works for you, I'll release v0.3.3 with that fix (as well as the PIN length message fix).

**Comment by kai-tub** [comment]
Again, sorry for the delay :bow: 
Yes, it works great now!
`age-plugin-yubikey --list` + `generate` work without any issues and in the logs I can see that my card reader is ignored :)
Feel free to close the issues once the branch has been merged :)
